[WIP] Write a JSON Schema case study for the W3C Web of Things

jviotti opened the pull request:
Fixes: https://github.com/json-schema-org/community/issues/221
Fixes: https://github.com/w3c/wot-marketing/issues/329
Signed-off-by: Juan Cruz Viotti <jv@jviotti.com>

jviotti commented:
Hi @Relequestual and @egekorkan,

I'm opening this PR with what we have so far so that we can collaborate and get it out soon! 

Some action items:

- [ ] We need a cover picture for the case study
- [ ] We need a round for review to see if there is anything we should change
- [ ] We need sign-off from the W3C WoT that this is fine to publish and no more corrections are needed 
- [ ] We need to set the right publication date for the article

Thanks a lot!

jviotti commented:
It doesn't seem like we will get a deployment preview of this branch for some reason. I'm attaching a PDF of how this looks like for me.

You can see it by running the following commands:

```
$ git clone https://github.com/json-schema-org/blog
$ cd blog
$ yarn
$ yarn dev
# Open this post at http://localhost:8080/blog/posts/w3c-wot-case-study
```

[How the W3C Web of Things brings JSON Schema to the Internet of Things | JSON Schema Blog.pdf](https://github.com/json-schema-org/blog/files/9366095/How.the.W3C.Web.of.Things.brings.JSON.Schema.to.the.Internet.of.Things.JSON.Schema.Blog.pdf)

egekorkan commented:
Hi @jviotti and @Relequestual ! Thanks for taking our input and turning it into a text that can be already read quite well. I can provide feedback, should I do it on the GitHub review or the PDF annotations? I would prefer GitHub option but I am ok for both. Regarding other points:
- Cover picture: We have the W3C WoT Logos at https://www.w3.org/WoT/IG/wiki/Web_of_Things_Logos but I am not sure if they would be good cover pictures. Otherwise, we have a cover picture at https://twitter.com/W3C_WoT . Another alternative would be that I can make a "WoT ❤️ JSON Schema" image using both logos?
- Regarding review process: As you have mentioned, we will need to give OK to publish, which is not a problem, the group has agreed that we want to do this. Just regarding the timeline, currently there will be some delays since I (and others too) am going on holidays for 2 weeks and then there are W3C events ([W3C TPAC in Vancouver](https://www.w3.org/2022/09/TPAC/)). I think that we can get it done before end of September though.

jviotti commented:
@egekorkan Awesome! Please post any comments on GitHub. The PDF is just a rendering on my local environment to simplify showing it to you :) 

> Cover picture: We have the W3C WoT Logos at https://www.w3.org/WoT/IG/wiki/Web_of_Things_Logos but I am not sure if they would be good cover pictures. Otherwise, we have a cover picture at https://twitter.com/W3C_WoT . Another alternative would be that I can make a "WoT ❤️ JSON Schema" image using both logos?

The Twitter cover looks pretty neat IMO, as it's more hardware-ish.

egekorkan commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,127 @@
+  - name: Juan Cruz Viotti
+    photo: /img/avatars/jviotti.webp
+    link: https://twitter.com/jviottidc
+excerpt: "Using JSON Schema at the W3C Web of Things to counter the fragmentation of the Internet of Things"
```
actually "countering the fragmentation" is something we want to avoid since it starts with a negative tone. I would propose (from our input document):

> Using JSON Schema at the W3C Web of Things to create an interoperability layer so that different IoT platforms, protocols and standards can operate together.

I hope it is not too late to fit into the small part that is under the title

jviotti replied:
Makes sense! Done

egekorkan commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,127 @@
+To solve these problems, the W3C Web of Things works on providing standardized
+building blocks that make use of JSON Schema.
+
+JSON Schema is used for validating abstractions of physical entities called
```
abstraction of a physical entity can be understood differently by different people. Not sure of the best replacement but maybe "description of network-facing capabilities of physical entities"

jviotti replied:
Yeah, I think that reads well. Done

egekorkan commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,127 @@
+[Thing Descriptions][thing-description], and to model and describe data sent by
+Internet of Things consumers and producers in a multi-protocol manner.
+
+We are on-board with JSON Schema since Draft 4, and even the first draft
```
Shouldn't we use another subject than "we"? There are other places with it as well.

jviotti replied:
True. Other case studies use "We" only in quotations. I'll fix it everywhere

jviotti replied:
Done

jviotti commented:
I have added WoT's Twitter cover to the post. I think it looks pretty good:

<img width="1033" alt="Screen Shot 2022-08-18 at 12 49 10" src="https://user-images.githubusercontent.com/2192773/185450778-eeda37f9-6870-409f-a81c-44dbc4a773c8.png">

jviotti commented:
I have addressed all your comments @egekorkan . Please give it another round of review! I also added a "thank you" note at the end like for other case studies.

egekorkan commented:
Thank you @jviotti ! Can you generate a new PDF so that I can send it for reviews in the WoT group?

handrews commented:
I just want to say I'm so happy to see this.  I enjoyed meeting the WoT folks at their conference in Santa Clara several years ago, when we had just started to revive work on the JSON Schema spec.  I'm really glad to see that the work we've done since then is of interest, and particularly excited to see how JSON Schema vocabularies can fit with Semantic Web approaches.

I never had as much time to keep up with the WoT work as I would have liked, but a better way to integrate with JSON LD/Semantic Web has always been something I'd love to see.  I prioritized OAS integration in part because they had a major revision timed right for 2020-12, but the Thing Description has always been another major JSON Schema use case I kept in mind.

egekorkan commented:
> I just want to say I'm so happy to see this. I enjoyed meeting the WoT folks at their conference in Santa Clara several years ago, when we had just started to revive work on the JSON Schema spec. I'm really glad to see that the work we've done since then is of interest, and particularly excited to see how JSON Schema vocabularies can fit with Semantic Web approaches.
> 
> I never had as much time to keep up with the WoT work as I would have liked, but a better way to integrate with JSON LD/Semantic Web has always been something I'd love to see. I prioritized OAS integration in part because they had a major revision timed right for 2020-12, but the Thing Description has always been another major JSON Schema use case I kept in mind.

Thank you for all your support over the years! Maybe @sebastiankb can remember more from Santa Clara event but having one the authors of JSON Schema since the beginning was a big support :)

jviotti commented:
@egekorkan Here you go!

[How the W3C Web of Things brings JSON Schema to the Internet of Things | JSON Schema Blog.pdf](https://github.com/json-schema-org/blog/files/9378105/How.the.W3C.Web.of.Things.brings.JSON.Schema.to.the.Internet.of.Things.JSON.Schema.Blog.pdf)

koalie commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -38,17 +38,19 @@ of physical entities called [Thing Descriptions][thing-description], and to
-becoming more and more consistent across implementations.  For example,
-vocabularies are very promising and we plan to explore them further in
-conjunction with Semantic Web technologies.
+The W3C Web of Things are on-board with JSON Schema since Draft 4, and even the
```
s/are on-board/specifications are on board/

koalie commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -38,17 +38,19 @@ of physical entities called [Thing Descriptions][thing-description], and to
-vocabularies are very promising and we plan to explore them further in
-conjunction with Semantic Web technologies.
+The W3C Web of Things are on-board with JSON Schema since Draft 4, and even the
+first draft versions were quite enough to model data models of Internet of
```
s/were quite enough to model/already catered to/

koalie commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -38,17 +38,19 @@ of physical entities called [Thing Descriptions][thing-description], and to
-conjunction with Semantic Web technologies.
+The W3C Web of Things are on-board with JSON Schema since Draft 4, and even the
+first draft versions were quite enough to model data models of Internet of
+Things devices.  "Currently, we are on Draft 7 and expect to move to 2020-12 or
```
What does "2020-12" mean? How does it relate to draft numbers which appear earlier in the sentence? I'm afraid it may be meaningless otherwise to people outside of the group.

handrews replied:
It might be good to include a link to https://json-schema.org/specification-links.html (which has a "Understanding draft names and numbers" section, because it's a surprisingly complex topic).

jviotti replied:
Make sense. Done!

handrews replied:
@koalie you asked how to resolve a comment- you should be able to see a "Resolve conversation" button below this comment reply.

koalie replied:
No I don't :( Can anyone else do that instead of me? I only see "review changes" at the same level as "conversation", "commits", "checks"; and no button at all below the comment reply.

koalie commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -38,17 +38,19 @@ of physical entities called [Thing Descriptions][thing-description], and to
+The W3C Web of Things are on-board with JSON Schema since Draft 4, and even the
+first draft versions were quite enough to model data models of Internet of
+Things devices.  "Currently, we are on Draft 7 and expect to move to 2020-12 or
+a newer one when starting our new charter in February 2023." - Ege Korkan
```
s/." - Ege Korkan/," said Ege Korkan, Research Scientist at Siemens and W3C Specifications editor.

koalie commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -38,17 +38,19 @@ of physical entities called [Thing Descriptions][thing-description], and to
+a newer one when starting our new charter in February 2023." - Ege Korkan
+
+"We observe JSON Schema becoming more and more stable over the years and
+Postman's support to the community gives us even more confidence on it." - Ege
```
s/on it." - Ege/in it," continued Ege/

koalie commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -38,17 +38,19 @@ of physical entities called [Thing Descriptions][thing-description], and to
+
+"We observe JSON Schema becoming more and more stable over the years and
+Postman's support to the community gives us even more confidence on it." - Ege
+Korkan 
```
s/Korkan/Korkan./

koalie commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -38,17 +38,19 @@ of physical entities called [Thing Descriptions][thing-description], and to
+features are becoming more and more consistent across implementations.  For
+example, "vocabularies are very promising and we plan to explore them further
+in conjunction with Semantic Web technologies." - Ege Korkan
 
```
I suggest you remove "- Ege Korkan", and instead write: For example, Ege Korkan added, "vocab[...]"

koalie commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -79,45 +81,47 @@ Descriptions.
+"Because we adopt JSON Schema, developers who are familiar with it have a
+smaller learning curve when learning the W3C Web of Things standard." - Ege
+Korkan
 
```
I suggest this rewrite:
Ege Korkan commented, "Because we use JSON Schema, developers who are familiar with it have a lesser learning curve when adopting the W3C Web of Things standard."

koalie commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -79,45 +81,47 @@ Descriptions.
 
 ## W3C Web of Things
 
 World Wide Web Consortium (W3C) is the standardization organization behind most
```
s/World/The World/

koalie commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -79,45 +81,47 @@ Descriptions.
 
 World Wide Web Consortium (W3C) is the standardization organization behind most
 of the standards used in the Web. We are the Web of Things Working and Interest
 Groups, who work on the standardization of Web of Things at the W3C.
```
s/at the W3C/at W3C/

koalie commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -79,45 +81,47 @@ Descriptions.
+together.
 
 The work has started as a discussion in the Web of Things Community Group in
 2013-2014.  It has proceed to an Interest Group in 2015 that has collected the
```
s/proceed/proceeded to become/

koalie commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -79,45 +81,47 @@ Descriptions.
-At the time of this writing, our Working Group consists of 96 participants
-representing 36 organizations and the our Interest Group consists of 131
+At the time of this writing, the Working Group consists of 96 participants
+representing 36 organizations and the the Interest Group consists of 131
```
s/the the/the/

sebastiankb commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,137 @@
+If you write your own Thing Description, you can programatically interact with
+your device by using the [Node.js Web of Things
+implementation](https://github.com/eclipse/thingweb.node-wot), or play with it
+through the [Node-RED](https://www.youtube.com/watch?v=oAcYbJ6P9bU)
```
you can also link to the node-red web of things plugin (based on node-wot): https://www.npmjs.com/package/node-red-contrib-web-of-things

jviotti commented:
Thanks a lot @koalie and @sebastiankb for the great comments. All of them should be resolved now!

sebastiankb commented:
> Thank you for all your support over the years! Maybe @sebastiankb can remember more from Santa Clara event but having one the authors of JSON Schema since the beginning was a big support :)

Indeed, I can :-) 

@handrews thank you for your support over the years when we had questions about JSON Schema 

Hopefully there will be a chance to meet again soon.

jviotti commented:
Ping @Relequestual. I think we are mainly missing your review here

Citrullin commented:
> Also, given that the cover picture was given to us by @Citrullin, @Citrullin is it fine to use it here (copyrights etc.)?

Thank you for pinging me regarding this. I own the full copyright to the work. It was only supposed to be used on the twitter profile. 
@jviotti Please use following license and attribution

[Hello WoT](https://bind.systems/tags/web-of-things/) © 2022 by desertmonitor OÜ is licensed under CC BY-ND 4.0

Relequestual commented:
@jviotti Going to review this now =]
@egekorkan Can you "resolve" your unresolved review comments please? =]
@koalie Looks like you have one unresolved review comment. Can you resolve that too please? =]

Relequestual commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,141 @@
+
+JSON Schema has proven to be a foundational block for creating higher-level
+tooling that is specific to the W3C Web of Things. As a notable example, the
+community has developed a tool that makes use of JSON Schema definitions inside
```
Is it worth linking to the tool?

jviotti replied:
I don't have the link. @egekorkan is this public?

egekorkan replied:
Here it is https://github.com/tum-esi/testbench

Relequestual commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,141 @@
+
+## Key Impact Results
+
+The ultimate metric of success for a standard like the W3C Web of Things is
+adoption, and community adoption is not something you can just "engineer".  Ege
+Korkan commented, "because we adopt JSON Schema, developers who are familiar
+with it have a smaller learning curve when learning the W3C Web of Things
+standard."
```
Hey @egekorkan. Here I'm ideally looking at some kind of soundbite.
Would you be able to take some sort of educated guess as to how much time it saves people getting to grips with the IoT standard?

egekorkan replied:
I actually have some sort of numbers on this. During my PhD times, I had quite a bit of students (electrical engineering and masters level) and they had to get up to speed with WoT. A good student took 1 week full time to go through my basic tutorial at https://github.com/tum-esi/wot-tutorial . I think that 1 week is too much for many people since my tutorial goes through Linux and git basics, then some JSON Schema and Node.js before starting with WoT at all. If someone is already somewhat experienced in JSON Schema and some basic Javascript, it should be 2 days full time.

jviotti replied:
Nice stuff! I tried to incorporate that example here. Let me know what you think!

Relequestual reviewed (approved):
Two comments / suggestions, but otherwise good to go. Nothing blocking from my perspective.

koalie commented:
>[...] @koalie Looks like you have one unresolved review comment. Can you resolve that too please? =]

I have no idea how to do that, sorry :(

jviotti commented:
Copyright mention added!

Relequestual commented:
> > [...] @koalie Looks like you have one unresolved review comment. Can you resolve that too please? =]
> 
> I have no idea how to do that, sorry :(

Looks like it's resolved now 🤷‍♂️.
I did hard refresh in case it was some sort of lag before asking. Ah well.
Thanks for your effort on this 🙌

Citrullin commented:
> Copyright mention added!

Can you please add a link to https://bind.systems/tags/web-of-things/
On the text "Hello WoT", Thank you :)

jviotti commented:
@Citrullin For sure. Done!

egekorkan commented:
@Relequestual comment of "@egekorkan Can you "resolve" your unresolved review comments please? =]" I am not able resolve them in this view but they dont exist in the files tab anyways. Given that the same happened to Coralie, I guess this is fine. I am back from holidays and will reply to the other comments now.

egekorkan reviewed (commented):
Here are my answers. Also, thanks to everyone for taking time in this, it was a nice collaboration :)

jviotti commented:
@Relequestual @egekorkan All comments resolved. Please give it one final read. If it looks good, then we can schedule the post!

egekorkan commented:
I'm give it another read next week but we will need to pass it through the working group once again. We have a big W3C event next week so there will be some delays :/

jviotti commented:
@egekorkan Here we go!

[How the W3C Web of Things brings JSON Schema to the Internet of Things | JSON Schema Blog.pdf](https://github.com/json-schema-org/blog/files/9548356/How.the.W3C.Web.of.Things.brings.JSON.Schema.to.the.Internet.of.Things.JSON.Schema.Blog.pdf)

egekorkan commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,149 @@
+and try them out on the [Thing Description
+Playground](http://plugfest.thingweb.io/playground/) app.
+
+Thank you to Ege Korkan, Web of Things Researcher at TU Munich, and the W3C Web
```
```suggestion
Thank you to Ege Korkan, Web of Things Researcher at Siemens AG, Munich, the W3C Web
```

egekorkan commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,149 @@
+Playground](http://plugfest.thingweb.io/playground/) app.
+
+Thank you to Ege Korkan, Web of Things Researcher at TU Munich, and the W3C Web
+of Things, for allowing and enabling us to share this case study with you.
```
```suggestion
of Things Working and Interest Groups and Coralie Mercier from the W3C Marketing and Communications Team for allowing and enabling us to share this case study with you.
```

egekorkan commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,149 @@
+collected the use cases and defined the standards to be worked on. Since 2016,
+the working group is working on different standards on the Web of Things with
+the first publications of the Thing Description and Architecture
+recommendations in 2019.
```
```suggestion
recommendations in 2019 and more that will be published by the end of this year.
```

egekorkan commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,149 @@
+
+We invite you to leverage your existing JSON Schema knowledge to play with the
+Internet of Things through the use of the W3C Web of Things specifications and
+tools. You might already have a smart device next to you that does not come
```
```suggestion
tools. You might already have a smart device next to you, just that it does not come
```

egekorkan reviewed (changes requested):
I have given it another read since forgetting about the text definitely helps ;) I will do another read next week

jviotti commented:
Added!

egekorkan commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,152 @@
+
+JSON Schema is delivering exciting features, and more importantly these
+features are becoming more and more consistent across implementations.  For
+example, Ege Korkan added, "vocabularies are very promising and we plan to
```
```suggestion
example, Ege Korkan added, "JSON Schema vocabularies are very promising and we plan to
```

better to be specific since vocabulary is a very commonly used term in WoT and Semantic Web. If you prefer another way to say that, I would be also fine :)

egekorkan commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,152 @@
+
+The ultimate metric of success for a standard like the W3C Web of Things is
+adoption, and community adoption is not something you can just "engineer".  Ege
+Korkan commented, "because we adopt JSON Schema, developers who are familiar
```
```suggestion
Korkan commented, "since we adopt JSON Schema, developers who are familiar
```

sounds better to me when this sentence is considered on its own

egekorkan commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,152 @@
+"During my PhD times, my MSc engineering students had to become familiar with
+the W3C Web of Things. These students typically took 1 week to go through my
+[tutorial](https://github.com/tum-esi/wot-tutorial). If they knew JSON Schema,
+that 1 week could become 2 days", said Ege Korkan.
```
```suggestion
that 1 week would become 2 days", said Ege Korkan.
```

egekorkan reviewed (changes requested):
Here is my last feedback after reading for one last time. We did not receive any further feedback from the WG so this can be published once these small changes are done.

jviotti commented:
Revisions accepted!

Relequestual commented:
I'm going to give it a final read through this week! =D

Relequestual commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,152 @@
+
+## W3C Web of Things
+
+The World Wide Web Consortium (W3C) is the standardization organization behind
+most of the standards used in the Web. We are the Web of Things Working and
+Interest Groups, who work on the standardization of Web of Things at W3C.
```
```suggestion
The World Wide Web Consortium (W3C) is the standardization organization behind
most of the standards used in the Web. The W3C Web of Things Working and
Interest Groups work on the standardization of Web of Things at W3C.
```

Missed one

Relequestual commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,152 @@
+different Internet of Things platforms, protocols and standards can operate
+together.
+
+The work has started as a discussion in the Web of Things Community Group in
+2013-2014.  It has proceeded to become an Interest Group in 2015 that has
+collected the use cases and defined the standards to be worked on. Since 2016,
+the working group is working on different standards on the Web of Things with
+the first publications of the Thing Description and Architecture
+recommendations in 2019 and more that will be published by the end of this
+year.
```
```suggestion
The work started as a discussion in the Web of Things Community Group in
2013-2014.  It proceeded to become an Interest Group in 2015 that
collected the use cases and defined the standards to be worked on. Since 2016,
the working group is working on different standards on the Web of Things with
the first publications of the Thing Description and Architecture
recommendations in 2019 and more that will be published by the end of this
year.
```

Minor gramatical corrections.

Relequestual commented on pages/posts/w3c-wot-case-study.md:
```diff
@@ -0,0 +1,152 @@
+
+## Getting Started
+
+We invite you to leverage your existing JSON Schema knowledge to play with the
```
```suggestion
The W3C Web of Things Working Group invite you to leverage your existing JSON Schema knowledge to play with the
```

egekorkan replied:
```suggestion
The W3C Web of Things Working Group invites you to leverage your existing JSON Schema knowledge to play with the
```

jviotti replied:
Done!

Relequestual reviewed (changes requested):
A few minor fixes, but the content is all good by me.
After tose are resolve, I say it's good to publish.

@egekorkan did you want to know when it **will** be published or just when it's published?
I'll probably merge in UK time, although anyone else is free to merge when my outstanding comments are resolved.

egekorkan commented:
Thanks @Relequestual , I think the changes are good, just that there was a missing s for the last one. I cannot suggest changes on a suggested change so I had to create a new one.

> @egekorkan did you want to know when it will be published or just when it's published?

Just when it is published :) I am not sure if the merge of this PR means publication, if it does, I will get the notification for sure :)

jviotti commented:
All done! @Relequestual Don't forget to update the post date before publishing though
